runtime/Cpp/runtime/src/atn/ArrayPredictionContext.cpp

```diff
@@ -22,6 +22,12 @@ namespace {
   }
 
   bool predictionContextEqual(const Ref<const PredictionContext> &lhs, const Ref<const PredictionContext> &rhs) {
+    if (lhs == nullptr && rhs == nullptr)
+        return true;
+    if (lhs == nullptr && !(rhs == nullptr))
+        return false;
+    if (!(lhs == nullptr) && rhs == nullptr)
+        return false;
```

Comment:
Please simplify:

```suggestion
    if (lhs == nullptr)
        return rhs == nullptr;
    else if (rhs == nullptr)
        return false;
```

Reply:
Sure.

But, I bet there are other places like this. For example, these two statements are very similar to this bug, with parameters of type `Ref<> &`.

* https://github.com/antlr/antlr4/blob/aa1f1f12a846846fcb78544bdb2ad135471376db/runtime/Cpp/runtime/src/atn/ATNConfig.h#L39
* https://github.com/antlr/antlr4/blob/aa1f1f12a846846fcb78544bdb2ad135471376db/runtime/Cpp/runtime/src/atn/PredictionContextCache.cpp#L55

People make changes that "fix" a tiny little problem, but don't read the damn code. @parrt what about these occurrences?

Reply:
Yep, that's why I'm opposed to refactoring for refactoring sake. It can introduce all sorts of subtle bugs.

We have to be careful here for example because the real solution is that context objects should never be null... There is a sentinel that indicates empty stack. That is why I didn't simply add the null checks... There's a larger bug and we shouldn't paint over it.

Reply:
> Yep, that's why I'm opposed to refactoring for refactoring sake. It can introduce all sorts of subtle bugs.
> 
> We have to be careful here for example because the real solution is that context objects should never be null... There is a sentinel that indicates empty stack. That is why I didn't simply add the null checks... There's a larger bug and we shouldn't paint over it.

Then, these should be at least assert()'s so that it crashes with a message.

Is there a method that anyone wrote that checks the integrity of the data structure so I can percolate the check around and snuff this bug out? Someone is creating an array with a null pointer at the end of the array.

Reply:
That's what I was hoping to do with the Data dump from the ATN simulator... We could use it to compare how the different targets work. 

BTW, here is the start of me trying to fix C++ by refracturing the complex code: https://github.com/antlr/antlr4/pull/3863

I gave up trying to debug when a print statement exploded all over my screen with compile errors.

Here is my attempt to start a tracer for ATN: https://github.com/antlr/antlr4/pull/3817

Reply:
Actually, I think your assumption that there cannot be null parents in an array of PredicateContext is wrong--or some of the ports are doing it wrong.

I wrote a Check() function that checks the integrity of the ArrayPredictionContext and it led to this code. This code creates a "parents" array containing a null pointer.

* Cpp:
https://github.com/antlr/antlr4/blob/16cfc6b122b184d1e412c7e175d4e12370b0b99d/runtime/Cpp/runtime/src/atn/PredictionContext.cpp#L320

Compare this with CSharp, Java, PHP, and Python3.

* CSharp:
https://github.com/antlr/antlr4/blob/16cfc6b122b184d1e412c7e175d4e12370b0b99d/runtime/CSharp/src/Atn/PredictionContext.cs#L391

* Java:
https://github.com/antlr/antlr4/blob/16cfc6b122b184d1e412c7e175d4e12370b0b99d/runtime/Java/src/org/antlr/v4/runtime/atn/PredictionContext.java#L324

* PHP:
https://github.com/antlr/antlr-php-runtime/blob/64fd1e2e8ca6fd80b57f53a0729e74b984bd767f/src/PredictionContexts/PredictionContext.php#L371

* Python3:
https://github.com/antlr/antlr4/blob/16cfc6b122b184d1e412c7e175d4e12370b0b99d/runtime/Python3/src/antlr4/PredictionContext.py#L420

For Dart, null pointers are disallowed (I think??), so the code looks very different.
* Dart:
https://github.com/antlr/antlr4/blob/16cfc6b122b184d1e412c7e175d4e12370b0b99d/runtime/Dart/lib/src/prediction_context.dart#L247

For Go, I think it too was written without using null pointers:
* Go:
https://github.com/antlr/antlr4/blob/16cfc6b122b184d1e412c7e175d4e12370b0b99d/runtime/Go/antlr/prediction_context.go#L536

Reply:
Well, dang. You're right. Hmm...yeah, touching the ATN simulation is really risky as I don't remember the details well. :(

Reply:
Yes, I can relate to forgetting details. It definitely is concerning that a couple of the targets don't have nulls in the parents array. I think to be "safe", I should test this against the regression tests of grammars-v4 for Cpp.

Reply:
I would love to do a comparison of the java target against other targets. I have good trust in java behavior.